compiler/src/dotty/tools/backend/jvm/BCodeSkelBuilder.scala

```diff
@@ -106,14 +109,82 @@ trait BCodeSkelBuilder extends BCodeHelpers {
 
       initJClass(cnode)
 
-      val methodSymbols = for (f <- cd.symbol.info.decls.toList if f.is(Method) && f.isTerm && !f.is(Module)) yield f
-      val hasStaticCtor = methodSymbols exists (_.isStaticConstructor)
-      if (!hasStaticCtor) {
-        // but needs one ...
-        if (isCZStaticModule || isCZParcelable) {
-          fabricateStaticInit()
+      val cd = if (isCZStaticModule) {
+        // Move statements from the primary constructor following the superclass constructor call to
+        // a newly synthesised tree representing the "<clinit>", which also assigns the MODULE$ field.
```

Comment:
Do you think this could be done in a mini-phase just before the backend?

Reply:
> Do you think this could be done in a mini-phase just before the backend?

Actually that might not be a goo idea if it interfers with scala.js/native like with https://github.com/lampepfl/dotty/pull/8652/commits/08a1ca0e6581ca997115ce1d483efa20d61427a4#r456815962, wdyt @sjrd ?

Reply:
Ah yes, moving that before the back-end would make things significantly more complicated for Scala.js and Native. Both have a first-class concept of `ModuleClass`, and encode the singleton pattern of `MODULE$` in a completely different way. In the Scala.js IR, there even isn't anything that corresponds to `MODULE$` per se.

So if you move this before the back-end, then Scala.js and Scala Native will have to reverse all the effects of this transformation, which would be counter-productive.